src/Microsoft.OData.Core/UriParser/SemanticAst/ODataPath.cs

```diff
@@ -77,7 +77,7 @@ public ODataPathSegment FirstSegment
         {
             get
             {
-                return this.segments.FirstOrDefault();

```

Comment:
>FirstOrDefault [](http://example.com/codeflow?start=37&length=14)

From the .NET Source code, if it's IList<T>, so, it returns the first one in the List directly?

private static TSource? TryGetFirst<TSource>(this IEnumerable<TSource> source, out bool found)
        {
            if (source == null)
            {
                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.source);
            }
 
            if (source is IPartition<TSource> partition)
            {
                return partition.TryGetFirst(out found);
            }
 
            if (source is IList<TSource> list)
            {
                if (list.Count > 0)
                {
                    found = true;
                    return list[0];
                }
            }
            else
            {
                using (IEnumerator<TSource> e = source.GetEnumerator())
                {
                    if (e.MoveNext())
                    {
                        found = true;
                        return e.Current;
                    }
                }
            }
 
            found = false;
            return default;
        }


The above has a bad perf?

Reply:
The .NET does that optimization for `FirstOrDefault` and `LastOrDefault`, however if I we already know this it's a list, then we can do it directly and avoid the extra layers of indirection that `FirstOrDefault` does, for example we avoid the casts (`source is IPartition<TSource> and `source is IList<TSource>`) and the extra method call. From the attached issue, you can see some CPU usage from `Enumerable.TryGetLast` that I think is worth optimizing if it's a low-hanging fruit.